src/coreclr/jit/morph.cpp

```diff
@@ -14738,6 +14714,46 @@ GenTree* Compiler::fgMorphModToSubMulDiv(GenTreeOp* tree)
     return sub;
 }
 
+//------------------------------------------------------------------------
+// fgMorphUModToAndSub: Transform a % b into the equivalent a & (b - 1).
+// '%' must be unsigned (GT_UMOD).
+// 'a' and 'b' must be integers.
+// 'b' must be a constant and a power of two.
+//
+// Arguments:
+//    tree - The GT_UMOD tree to morph
+//
+// Returns:
+//    The morphed tree
+//
+// Notes:
+//    This is more optimized than calling fgMorphModToSubMulDiv.
+//
+GenTree* Compiler::fgMorphUModToAndSub(GenTreeOp* tree)
+{
+    JITDUMP("\nMorphing UMOD [%06u] to And/Sub\n", dspTreeID(tree));
+
+    if (tree->OperGet() != GT_UMOD)
+    {
+        noway_assert(!"Illegal gtOper in fgMorphUModToAndSub");
```

Comment:
just curious why it can't be just `assert(tree->OperIs(GT_UMOD));` ? do you expect this assumption to fail with some probability? (that's when we use `noway_assert` in order to de-opt)

Reply:
It can be, I seemed to be following existing code doing it this way - I can change it as it's clearer.

Reply:
`assert()` would be appropriate here.